File: algorithms/code/leetcode/lc078_subsets/lc078_subsets.py
# https://leetcode.com/problems/subsets/

import unittest


class Solution:
    def __subsets(
        self,
        nums: list[int],
        offset: int,
        current: list[int],
        listed: list[list[int]],
    ) -> None:
        listed.append(current.copy())
        for i in range(offset, len(nums)):
            current.append(nums[i])
            self.__subsets(nums, i + 1, current, listed)
            current.pop()

    def subsets(self, nums: list[int]) -> list[list[int]]:
        nums.sort()
        listed: list[list[int]] = []
        self.__subsets(nums, 0, [], listed)
        return listed


class TestCode(unittest.TestCase):
    def __test(
        self, expected: list[list[int]], result: list[list[int]]
    ) -> None:
        result = sorted(result)
        expected = sorted(expected)
        assert len(expected) == len(result)
        for i, expected_i in enumerate(expected):
            assert len(expected_i) == len(result[i])
            result_i = sorted(result[i])
            expected_i = sorted(expected_i)
            assert expected_i == result_i

    def test_1_2_3(self) -> None:
        listed = Solution().subsets([1, 2, 3])
        expected = [[], [1], [2], [3], [1, 2], [1, 3], [2, 3], [1, 2, 3]]
        self.__test(expected, listed)

File: algorithms/code/leetcode/lc078_subsets/lc078_subsets.rs
// https://leetcode.com/problems/subsets/
// #medium

pub struct Solution;

impl Solution {
    fn subsets_internal(
        nums: &[i32],
        offset: usize,
        current: &mut Vec<i32>,
        list: &mut Vec<Vec<i32>>,
    ) {
        list.push(current.clone());
        for i in offset..nums.len() {
            current.push(nums[i]);
            Self::subsets_internal(nums, i + 1, current, list);
            current.pop();
        }
    }

    pub fn subsets(mut nums: Vec<i32>) -> Vec<Vec<i32>> {
        nums.sort_unstable();
        let mut list = Vec::new();
        Self::subsets_internal(&nums, 0, &mut Vec::new(), &mut list);
        list
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_123() {
        let mut list = Solution::subsets(vec![1, 2, 3]);
        let mut expected = vec![
            vec![],
            vec![1],
            vec![2],
            vec![3],
            vec![1, 2],
            vec![1, 3],
            vec![2, 3],
            vec![1, 2, 3],
        ];
        list.sort_unstable();
        expected.sort_unstable();
        assert_eq!(list.len(), expected.len());
        for i in 0..expected.len() {
            let mut sorted_list = list[i].clone();
            sorted_list.sort_unstable();
            let mut sorted_expected = expected[i].clone();
            sorted_expected.sort_unstable();
            assert_eq!(sorted_list, sorted_expected);
        }
    }
}
